src/Microsoft.Health.Fhir.Shared.Core/Features/Validation/ProfileValidator.cs

```diff
@@ -4,43 +4,125 @@
+using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
 using Microsoft.Health.Fhir.Core.Configs;
 using Microsoft.Health.Fhir.Core.Models;
+using Polly;
+using Task = System.Threading.Tasks.Task;
 
 namespace Microsoft.Health.Fhir.Core.Features.Validation
 {
-    public class ProfileValidator : IProfileValidator
+    public class ProfileValidator : IProfileValidator, IHostedService
     {
-        private readonly IResourceResolver _resolver;
+        private readonly ILogger<ProfileValidator> _logger;
+        private readonly IProvideProfilesForValidation _profilesResolver;
+        private readonly Lazy<IResourceResolver> _resolver;
+        private readonly CoreFeatureConfiguration _coreConfig;
+        private readonly ValidateOperationConfiguration _options;
 
-        public ProfileValidator(IProvideProfilesForValidation profilesResolver, IOptions<ValidateOperationConfiguration> options)
+        public ProfileValidator(
+            IProvideProfilesForValidation profilesResolver,
+            IOptions<CoreFeatureConfiguration> coreFeatureConfiguration,
+            IOptions<ValidateOperationConfiguration> options,
+            ILogger<ProfileValidator> logger)
         {
-            EnsureArg.IsNotNull(profilesResolver, nameof(profilesResolver));
-            EnsureArg.IsNotNull(options?.Value, nameof(options));
+            _profilesResolver = EnsureArg.IsNotNull(profilesResolver, nameof(profilesResolver));
+            _options = EnsureArg.IsNotNull(options?.Value, nameof(options));
+            _coreConfig = EnsureArg.IsNotNull(coreFeatureConfiguration?.Value, nameof(options));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
 
+            _resolver = new Lazy<IResourceResolver>(
+                () => Policy.Handle<Exception>()
+                    .WaitAndRetry(5, i => TimeSpan.FromSeconds(i * 2))
+                    .Execute(Initialize),
+                LazyThreadSafetyMode.ExecutionAndPublication);
+        }
+
+        private IResourceResolver Initialize()
+        {
             try
             {
-                _resolver = new MultiResolver(new CachedResolver(ZipSource.CreateValidationSource(), options.Value.CacheDurationInSeconds), profilesResolver);
+                (string Server, string CorePackageName, string ExpansionsPackageName) packageVariables = GetFhirPackageVariables();
+
+                var packagesToInclude = new HashSet<string>();
+
+                if (_coreConfig.PackageConfiguration.IncludeDefaultPackages)
+                {
+                    packagesToInclude.Add(packageVariables.CorePackageName);
+                    packagesToInclude.Add(packageVariables.ExpansionsPackageName);
+
+                    // elements package name is not included in the reflection metadata, but we should try and include it
+                    var extensionsPackageName = packageVariables.ExpansionsPackageName.Replace("expansions", "elements", StringComparison.OrdinalIgnoreCase);
```

Comment:
Can we put this behind the `#if !R5` syntax since R5 doesn't have the extensions package?

Reply:
I thought we DO also want expansions of the built-in codes?